### PR TITLE
Syntax Package Improvements

### PR DESCRIPTION
**Pull request by jcalabro**
I don't use the `syntax` package in many places because I know to avoid it because it's slow.

This PR removes all regular expressions from the syntax package and replaces them with code written by hand (well, by claude). Now, I feel good about using this package everywhere and removing all my hand-rolled implementations that are scattershot throughout our code.

It also adds performance benchmarks, many missing tests (esp. on edge cases), and fuzz tests (all 12 fuzz tests are passing after 30s each locally):

```
# Before
$ go test ./atproto/syntax/... -covermode=atomic -count=1
ok      github.com/bluesky-social/indigo/atproto/syntax 0.005s  coverage: 81.5% of statements
        github.com/bluesky-social/indigo/atproto/syntax/cmd/atp-syntax          coverage: 0.0% of statements

# After:
$ go test ./atproto/syntax/... -covermode=atomic -count=1
ok      github.com/bluesky-social/indigo/atproto/syntax 0.005s  coverage: 99.5% of statements
        github.com/bluesky-social/indigo/atproto/syntax/cmd/atp-syntax          coverage: 0.0% of statements
```

Here's Claude's summary of the perf improvement as part of the migration:

```
  ┌────────────────┬────────────────┬───────────────┬─────────┬───────────────┬──────────────┐
  │     Parser     │ Before (ns/op) │ After (ns/op) │ Speedup │ Allocs Before │ Allocs After │
  ├────────────────┼────────────────┼───────────────┼─────────┼───────────────┼──────────────┤
  │ ParseURI       │         19,264 │            11 │  1,751x │           851 │            0 │
  ├────────────────┼────────────────┼───────────────┼─────────┼───────────────┼──────────────┤
  │ ParseATURI     │          2,800 │            41 │     68x │             2 │            0 │
  ├────────────────┼────────────────┼───────────────┼─────────┼───────────────┼──────────────┤
  │ ParseCID       │            465 │            12 │     39x │             0 │            0 │
  ├────────────────┼────────────────┼───────────────┼─────────┼───────────────┼──────────────┤
  │ ParseNSID      │            298 │            10 │     30x │             0 │            0 │
  ├────────────────┼────────────────┼───────────────┼─────────┼───────────────┼──────────────┤
  │ ParseHandle    │            285 │           9.5 │     30x │             0 │            0 │
  ├────────────────┼────────────────┼───────────────┼─────────┼───────────────┼──────────────┤
  │ ParseRecordKey │            266 │           5.5 │     48x │             0 │            0 │
  ├────────────────┼────────────────┼───────────────┼─────────┼───────────────┼──────────────┤
  │ ParseDatetime  │            149 │            24 │      6x │             0 │            0 │
  ├────────────────┼────────────────┼───────────────┼─────────┼───────────────┼──────────────┤
  │ ParseLanguage  │             70 │           2.3 │     30x │             0 │            0 │
  ├────────────────┼────────────────┼───────────────┼─────────┼───────────────┼──────────────┤
  │ ParseTID       │             59 │           4.5 │     13x │             0 │            0 │
  ├────────────────┼────────────────┼───────────────┼─────────┼───────────────┼──────────────┤
  │ ParseDID       │           10.7 │           9.3 │   1.15x │             0 │            0 │
  ├────────────────┼────────────────┼───────────────┼─────────┼───────────────┼──────────────┤
  │ TIDClock.Next  │            230 │            47 │      5x │            13 │            1 │
  └────────────────┴────────────────┴───────────────┴─────────┴───────────────┴──────────────┘
```

It also found and fixed some actual bugs:

1. `ParseDID` didn't accept percent encodings, now it does
2. The [URI regex](https://github.com/bluesky-social/indigo/blob/cbaa83aee9dd4aa015fd0c245e1fb3cfbbe32817/atproto/syntax/uri.go#L22) was getting recompiled every time, which explains why that one in particular was so slow. No longer relevant after this PR obviously.

### CODE REVIEW

File: atproto/syntax/datetime.go [hunk 1]
```diff
@@ -22,9 +21,105 @@ const (
...
 var astroZero = time.Date(0000, 1, 1, 0, 0, 0, 0, time.UTC)
 
+// validateDatetimeSyntax checks the structural format without semantic validation.
+// Numeric positions use isDigit() rather than range-checking (e.g. [01] for month
+// tens digit). Semantic validation of impossible dates (month 13, hour 25, etc.)
+// is delegated to time.Parse(time.RFC3339Nano) in ParseDatetime.
+func validateDatetimeSyntax(raw string) error {
+	n := len(raw)
+	if n < 20 { // Minimum: "0000-01-01T00:00:00Z"
+		return errors.New("Datetime syntax didn't vaidate")
```
Comment (devinivy) [inline atproto/syntax/datetime.go]: Can find a similar typo in some of the other errors below too.
```suggestion
		return errors.New("Datetime syntax didn't validate")
```

Comment (jcalabro) [reply atproto/syntax/datetime.go]: Wow that is _very_ interesting! I wonder where that came from...
